yt_dlp/extractor/substack.py

```diff
@@ -0,0 +1,103 @@
+        'md5': 'fd3c07077b02444ff0130715b5f632bb',
+        'info_dict': {
+            'id': '47368578',
+            'ext': 'mp4',
+            'title': 'Mussels with Black Bean Sauce: Recipe of the Week #7',
+            'description': 'md5:b96234a2906c7d854d5229818d889515',
+            'thumbnail': 'md5:e30bfaa9da40e82aa62354263a9dd232',
+            'uploader': "Andrew Zimmern's Spilled Milk ",
+            'uploader_id': '7919490',
+        }
+    }]
+
+    @classmethod
+    def _extract_url(cls, webpage, url):
+        if not re.search(r'<script[^>]+src=["\']https://substackcdn.com/[^"\']+\.js', webpage):
+            return
+
+        mobj = re.search(r'{[^}]*["\']subdomain["\']\s*:\s*["\'](?P<subdomain>[^"]+)', webpage)
+        if mobj:
+            parsed = urllib.parse.urlparse(url)
+            return parsed._replace(netloc=f'{mobj.group("subdomain")}.substack.com').geturl()
+
+    def _extract_video_formats(self, video_id, username):
+        formats, subtitles = [], {}
+        for video_format in ('hls', 'mp4'):
+            video_url = f'https://{username}.substack.com/api/v1/video/upload/{video_id}/src?type={video_format}'
+
+            if video_format == 'hls':
+                fmts, subs = self._extract_m3u8_formats_and_subtitles(video_url, video_id, 'mp4', fatal=False)
+                formats.extend(fmts)
+                self._merge_subtitles(subs, target=subtitles)
+            else:
+                formats.append({
+                    'url': video_url,
+                    'ext': video_format,
+                })
+
+        return formats, subtitles
+
+    def _real_extract(self, url):
+        display_id, username = self._match_valid_url(url).group('id', 'username')
+        webpage = self._download_webpage(url, display_id)
+
+        webpage_info = self._search_json(r'<script[^>]*>\s*window\._preloads\s*=', webpage, 'preloads', display_id)
+
+        post_type = webpage_info['post']['type']
+        formats, subtitles = [], {}
+        if post_type == 'podcast':
+            formats, subtitles = [{'url': webpage_info['post']['podcast_url']}], {}
+        elif post_type == 'video':
+            formats, subtitles = self._extract_video_formats(webpage_info['post']['videoUpload']['id'], username)
+        else:
+            self.raise_no_formats(f'Page type "{post_type}" is not supported')
+
+        self._sort_formats(formats)
+        return {
+            'id': str(webpage_info['post']['id']),
+            'formats': formats,
+            'subtitles': subtitles,
+            'title': traverse_obj(webpage_info, ('post', 'title')),
+            'description': traverse_obj(webpage_info, ('post', 'description')),
+            'thumbnail': traverse_obj(webpage_info, ('post', 'cover_image')),
+            'uploader': traverse_obj(webpage_info, ('pub', 'name')),
+            'uploader_id': str_or_none(traverse_obj(webpage_info, ('pub', 'author_id'))),
```

Comment:
`.pub.name` identifies the newsletter, but `.pub.author_id` identifies the author. If you want to have ‘uploader’ mean ‘newsletter’ for this extractor, you should use `.pub.id` or `.post.publication_id` as `uploader_id`.

Also take note that some newsletters may have multiple authors. <https://www.mollymovieclub.com/p/interstellar> names two authors, but this extracts only one of them.

I remember there being some (proposal to add?) ‘crew’ credits metadata fields, maybe post authors (`.publishedBylines`) could be squeezed there.

Reply:
Should be good now

The closest thing that I found in some extractors is `cast`, could it work?

Reply:
Maybe. But I don’t insist on it.